### PR TITLE
Fix havocking of dependent assigns clause targets

### PR DESCRIPTION
**Pull request by remi-delmas-3000**
The first commit fixes a minor issue with history variables (allow history variables for constant and typecast expressions).

The second commit fixes missing declarations for return value variables introduced during replacement of function calls by contracts.

The third commit fixes a problem with assigns clause havocking when replacing calls by contracts.

When replacing a function call by its contract, we would directly havoc the targets of the assigns clause in an unspecified order. If the assigns clause contained dependent targets such as `assigns(s->ptr, *(s->ptr))`, havocking `s->ptr` before `*s->ptr` would lead to errors and incorrect results. 

We now introduce independent snapshot variables storing the address of each target and havoc them in a second time to avoid this dependency on the ordering of havoc instructions.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.cpp:570]: Is there a test that demonstrates the need for this?

File: regression/contracts/assigns_replace_08/test.desc
Comment (tautschnig) [inline regression/contracts/assigns_replace_08/test.desc]: I'm struggling to parse this - should it perhaps say "replaced *function* bar?"

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.h
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.h]: Forward declarations (`class namespacet;`, `class symbol_tablet;`) will do the trick here. (This is to reduce compilation times where possible.)

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.h
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.h]: This isn't used here.

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.h
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.h]: I think you'll only need `#include <util/expr.h>`, everything else can be forward-declared.

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp:61]: Please use `to_unary_expr(target).op()` (for a minimal amount of sanity checking).

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: Oh, we're really lacking a proper constructor for `minus_exprt` here. Adding this to my todo list.

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: Please move this into the `INVARIANT` to avoid a situation where `size` ends up as an unused value (when completely disabling invariants). That is: `INVARIANT(size_of_expr(target.type(), ns).has_value(), ...)`

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: Wouldn't `is_assignable` be true for something like `a[5]`, but taking the address thereof would not actually be the "base address" (or at least not how I'd understand that term).

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: For debugging purposes it would be helpful if all these had source locations that relate them in some way to the location a snapshot is taken at. Hmm, you're doing this after the fact. That's a bit confusing, in my opinion.

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: You could use `insert_before_swap` to avoid the `skip_program`. But then it probably doesn't make a whole lot of a difference.

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: `codet code(ID_havoc_object, {target_snapshot_var});` would avoid the two-step construction.

**Review by tautschnig — APPROVED** [review]

File: regression/contracts/assigns_replace_08/test.desc
Comment (feliperodri) [reply regression/contracts/assigns_replace_08/test.desc]: How about `Check that \*z is assignable (function bar replacement): `?

File: regression/contracts/assigns_replace_havoc_dependent_targets/vect.h
Comment (feliperodri) [inline regression/contracts/assigns_replace_havoc_dependent_targets/vect.h]: Why do you need to check whether `v->size` is smaller than `__CPROVER_max_malloc_size`? Haven't this problem been solved (see https://github.com/diffblue/cbmc/issues/6130)?

File: src/goto-instrument/contracts/contracts.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: Is there a issue to track this?

Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: nit:
```suggestion
    // Havoc all targets in the assigns clause.
    // TODO: handle local statics possibly touched by this function.
```

File: regression/contracts/assigns_replace_havoc_dependent_targets_fail/enforce.desc
Comment (feliperodri) [inline regression/contracts/assigns_replace_havoc_dependent_targets_fail/enforce.desc]: Is that the correct file name? I didn't find it.

File: regression/contracts/assigns_replace_havoc_dependent_targets_fail/vect.h
Comment (feliperodri) [inline regression/contracts/assigns_replace_havoc_dependent_targets_fail/vect.h:35]: See https://github.com/diffblue/cbmc/issues/6130.

Hasn't this problem been fixed? Why do you need `v->size <= __CPROVER_max_malloc_size` here?

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: This computes the start address of the thing that gets assigned. If contract assigns `a[5]` the base address of the thing that gets assigned is `&(a[5])`. If the contract says it assigns `__CPROVER_pointer_object(a[5])` then the base address will be the address pointed to by `a-5`.

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.h
Comment (feliperodri) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.h]: typo: `__CPROVER_assigns`

**Review by feliperodri — APPROVED** [review]
Only minor comments.

File: regression/contracts/assigns_replace_havoc_dependent_targets/vect.h
Comment (remi-delmas-3000) [reply regression/contracts/assigns_replace_havoc_dependent_targets/vect.h]: Whatever the behaviour of CBMC is when this bound is exceeded, I want to avoid it in this example.

File: regression/contracts/assigns_replace_havoc_dependent_targets_fail/vect.h
Comment (remi-delmas-3000) [reply regression/contracts/assigns_replace_havoc_dependent_targets_fail/vect.h:35]: Whatever the behaviour of CBMC is when this bound is exceeded, I want to avoid it in this example.

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp:570]: The tests added in this PR will fail without this.

Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp:570]: added specific tests

File: regression/contracts/assigns_replace_08/test.desc
Comment (remi-delmas-3000) [reply regression/contracts/assigns_replace_08/test.desc]: yes

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: fixed

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: fixed

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: it seems less natural, since after inserting the skip statement in dest, you would have to use a mix of insert_before_swap and skip_target++ or to insert instructions in reverse, or to create another program and insert_before_swap and then ++ the skip_target

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: fixed

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: yes

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (tautschnig) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: Ok!

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (tautschnig) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: I'd then find something along the lines of "build_address_of" a more natural function name, but this might be a matter of personal preference.

Comment (kroening) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: As a side, note that `address_of_exprt` has a convenience constructor that takes just the object, i.e., the above is equivalent to `address_of_exprt{target}`. The constructor will create the pointer type from the type of `target`.

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: Scratch my earlier comment, the following will actually do the right thing: `return minus_exprt{ptr, pointer_offset(ptr)};` (the type of the `minus_exprt` will be `ptr.type()` when constructed in this way).

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.h]: fixed

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.h]: fixed

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.h]: indeed this works

File: regression/contracts/assigns_replace_havoc_dependent_targets_fail/enforce.desc
Comment (remi-delmas-3000) [reply regression/contracts/assigns_replace_havoc_dependent_targets_fail/enforce.desc]: fixed

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.h]: fixed

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: thanks

File: src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp]: renamed to `build_address_of`

**Comment by remi-delmas-3000** [comment]
I took into account all pending comments, it should be good to go now.

**Comment by tautschnig** [comment]
Requires approval from @peterschrammel @martin-cs @chrisr-diffblue as a new file was added to `goto-instrument/Makefile`.

File: regression/contracts/history-constant/main.c
Comment (feliperodri) [inline regression/contracts/history-constant/main.c:1]: Why do we need this include?

File: regression/contracts/history-typecast/main.c
Comment (feliperodri) [inline regression/contracts/history-typecast/main.c:1]: Why do we need this include?

**Review by feliperodri — APPROVED** [review]
It needs code owner approval from @peterschrammel @chrisr-diffblue or @martin-cs.

**Review by chrisr-diffblue — APPROVED** [review]
Approving the Makefile changes. Nothing to add beyond the other detailed reviews.

**Pull request merged**
